test_conformance/subgroups/test_subgroup_non_uniform_vote.cpp

```diff
@@ -0,0 +1,303 @@
+                                "work group %d. Expected: %d Obtained: %d\n",
+                                i, j, k, tr, rr);
+                            return TEST_FAIL;
+                        }
+                    }
+                }
+            }
+
+            x += nw;
+            y += nw;
+            m += 4 * nw;
+        }
+
+        log_info("  sub_group_%s%s... passed\n",
+                 (operation == NonUniformVoteOp::elect) ? "" : "non_uniform_",
+                 operation_names(operation));
+        return TEST_PASS;
+    }
+};
+static const char *elect_source = R"(
+    __kernel void test_elect(const __global Type *in, __global int4 *xy, __global Type *out) {
+        int gid = get_global_id(0);
+        XY(xy,gid);
+        uint elect_work_item = 1 << (get_sub_group_local_id() % 32);
+            if (elect_work_item & WORK_ITEMS_MASK){
+                out[gid] = sub_group_elect();
+            }
+    }
+)";
+
+static const char *non_uniform_any_source = R"(
+    __kernel void test_non_uniform_any(const __global Type *in, __global int4 *xy, __global Type *out) {
+        int gid = get_global_id(0);
+        XY(xy,gid);
+        uint elect_work_item = 1 << (get_sub_group_local_id() % 32);
+            if (elect_work_item & WORK_ITEMS_MASK){
+                out[gid] = sub_group_non_uniform_any(in[gid]);
+            }
+    }
+)";
+
+static const char *non_uniform_all_source = R"(
+    __kernel void test_non_uniform_all(const __global Type *in, __global int4 *xy, __global Type *out) {
+        int gid = get_global_id(0);
+        XY(xy,gid);
+        uint elect_work_item = 1 << (get_sub_group_local_id() % 32);
+            if (elect_work_item & WORK_ITEMS_MASK){
+                out[gid] = sub_group_non_uniform_all(in[gid]);
+            }
+    }
+)";
+
+static const char *non_uniform_all_equal_source = R"(
+    __kernel void test_non_uniform_all_equal(const __global Type *in, __global int4 *xy, __global Type *out) {
+        int gid = get_global_id(0);
+        XY(xy,gid);
+        uint elect_work_item = 1 << (get_sub_group_local_id() % 32);
+            if (elect_work_item & WORK_ITEMS_MASK){
+                out[gid] = sub_group_non_uniform_all_equal(in[gid]);
+            }
+    }
+)";
+
+template <typename T> int run_vote_all_equal_for_type(RunTestForType rft)
+{
+    int error = rft.run_impl<T, VOTE<T, NonUniformVoteOp::all_equal>>(
+        "test_non_uniform_all_equal", non_uniform_all_equal_source);
+    return error;
+}
+}
+
+int test_subgroup_functions_non_uniform_vote(cl_device_id device,
+                                             cl_context context,
+                                             cl_command_queue queue,
+                                             int num_elements)
+{
+    std::vector<std::string> required_extensions = {
+        "cl_khr_subgroup_non_uniform_vote"
+    };
+
+    std::vector<uint32_t> masks{ 0xffffffff, 0x55aaaa55, 0x5555aaaa, 0xaaaa5555,
```

Comment:
The masks here are specific to a maximum subgroup size of 32. This is not appropriate. Please rewrite using a bitset of 128 bits, with one bit per work item for a maximum subgroup size of 128. Please use appropriate bit patterns to ensure reasonable coverage, preferably generated computationally.

The testing here lacks coverage of elect operations where the elected subgroup has a subgroup local ID greater than or equal to 32. The testing here does not ensure that any subgroup local ID (within the maximum subgroup size) is electable.

I have drawn special attention to elect operations above, because these are the operations for which the gap in coverage is most likely to hide an implementation bug. However, it is also not appropriate that the testing for sub_group_non_uniform_all, sub_group_non_uniform_all and sub_group_non_uniform_all_equal also uses masks specific to a maximum subgroup size of 32. Testing for these also does not cover cases where all work items with a subgroup local ID below 32 are inactive, and does not reasonably ensure that correct results are obtained when arbitrary sets of work items are active.

Note that for cl_khr_subgroup_non_uniform_vote, there is no direct reference to a maximum possible subgroup size of 128, but as other extensions in the family _do_ place a limit on the maximum subgroup size of 128 work items, it is reasonable to apply that same limit here.

Reply:
In my opinion this code is not limiting only to subgroup size == 32.
Please look at the lines where active workitems are chosen:
in verification function:
https://github.com/gwawiork/OpenCL-CTS/blob/extsub_2711_2020_all_changes/test_conformance/subgroups/test_subgroup_non_uniform_vote.cpp#L144
in kernel:
https://github.com/gwawiork/OpenCL-CTS/blob/extsub_2711_2020_all_changes/test_conformance/subgroups/test_subgroup_non_uniform_vote.cpp#L222
When subgroup_local_id is bigger than 32 operation `%` cause that calculated number will be always between 0 and 31 to match mask 32 bit size.

Reply:
I have read the code several times. The testing lacks coverage of elect operations where the elected subgroup has a subgroup local ID greater than or equal to 32.

If I am incorrect, please provide a single example of an instance in which the sub_group_elect() call returns a value of 32 or greater.